### ISSUE TITLE
Help with unit tests

### ISSUE THREAD

**Issue by codingwiththiago** [launchdarkly/flutter-client-sdk#10]
**Is your feature request related to a problem? Please describe.**
No.

**Describe the solution you'd like**
I would like to be able to mock the requests, so I can write some unit tests on the feature. Right now, all the necessary calls are done to `LDClient`, which cannot be instantiated, and therefore cannot be tested.

**Describe alternatives you've considered**
If there was any way to start the `LDClient` with a mock, and then control which answers we would get, that would be more than enough to test it properly.

**Comment by louis-launchdarkly** [comment]
Hello @codingwiththiago, 

Thanks for reaching out to us.

From what you have described, the [Test Data Source](https://docs.launchdarkly.com/sdk/features/test-data-sources) feature we built for other SDKs will allow you to test `LDClient` without a data source connecting to LaunchDarkly.

While we do not have an ETA for this feature for Flutter yet, this feature is in our backlog. We will let you know when there is an update.

**Comment by kinyoklion** [comment]
Hello @codingwiththiago,

In version 4.0+ the client is an instance and it exposes an interface so that it can be mocked.

Thank you,
Ryan

**Comment by danallen88** [comment]
> In version 4.0+ the client is an instance and it exposes an interface so that it can be mocked.

This works really well, and I appreciate the move away from singletons to make it easier to use dependency injection with wrapper classes for these clients. I am having some difficulty in dealing with `LDContext` being a final class, however.

In reading https://github.com/dart-lang/mockito/issues/635, the general recommendation (without changing the library) would be to wrap the plugin's class (in this case LDContext), but I'm not sure how that would work in the case of using the `identify` method.  Here's an example:

```dart
  Future<void> identify({
    required String userId,
    required String deviceId,
    required String practiceId,
  }) async {
    try {
      final context = buildContext(
        userId: userId,
        deviceId: deviceId,
        practiceId: practiceId,
      );

      final identifyResult =
          await _ldClient.identify(context).timeout(const Duration(seconds: 5));

      if (identifyResult is IdentifyError) {
        throw LaunchDarklyClientIdentificationException(identifyResult.error);
      }
    } catch (error, stackTrace) {
      Error.throwWithStackTrace(
        LaunchDarklyClientIdentificationException(error),
        stackTrace,
      );
    }

  /// {@template launch_darkly_plugin.build_context}
  /// Creates an [LDContext] to associate with LaunchDarkly, using the provided
  /// optional parameters.
  /// {@endtemplate}
  static LDContext buildContext({
    String? userId,
    String? deviceId,
    String? practiceId,
  }) {
    final contextBuilder = LDContextBuilder();

    if (userId?.isNotEmpty ?? false) {
      contextBuilder.kind(LaunchDarklyContext.user, userId);
    }

    if (deviceId?.isNotEmpty ?? false) {
      contextBuilder.kind(LaunchDarklyContext.device, deviceId);
    }

    if (practiceId?.isNotEmpty ?? false) {
      contextBuilder.kind(LaunchDarklyContext.practiceId, practiceId);
    }

    return contextBuilder.build();
  }
```

Let's say I wanted to test this method under the different circumstances where plugin's `_ldClient.identify` method returned different variations on an `IdentifyResult`.  My test code would look something like this:

```dart
class _MockLDClient extends Mock implements LDClient {}

void main() {
  group('LaunchDarklyClient', () {
    const deviceId = 'deviceId';
    const userId = 'userId';
    const practiceId = 'practiceId';

    late LaunchDarklyClient launchDarklyClient;
    late LDClient ldClient;
    late LDContext ldContext;

    setUp(() {
      ldClient = _MockLDClient();
      ldContext = LaunchDarklyClient.buildContext(
        userId: userId,
        deviceId: deviceId,
        practiceId: practiceId,
      );

      launchDarklyClient = LaunchDarklyClient(ldClient: ldClient);
    });

group('identify', () {
      setUp(() {
        when(
          () => ldClient.identify(ldContext),
        ).thenAnswer((_) => Future.value(IdentifyComplete()));
      });

      group('returns normally', () {
        test('when LDClient.identify returns IdentifyComplete', () async {
          when(
            () => ldClient.identify(ldContext),
          ).thenAnswer((_) => Future.value(IdentifyComplete()));

          expect(
            () => launchDarklyClient.identify(
              userId: userId,
              deviceId: deviceId,
              practiceId: practiceId,
            ),
            returnsNormally,
          );
        });

        test('when LDClient.identify returns IdentifySuperseded', () async {
          when(
            () => ldClient.identify(ldContext),
          ).thenAnswer((_) => Future.value(IdentifySuperseded()));

          expect(
            () => launchDarklyClient.identify(
              userId: userId,
              deviceId: deviceId,
              practiceId: practiceId,
            ),
            returnsNormally,
          );
        });
      });

      test(
        'calls ldClient.identify with userId, practiceId, and deviceId in '
        'context',
        () async {
          await launchDarklyClient.identify(
            userId: userId,
            deviceId: deviceId,
            practiceId: practiceId,
          );

          verify(
            () => ldClient.identify(ldContext),
          ).called(1);
        },
      );

      group('throws a LaunchDarklyClientIdentificationException', () {
        test('when LDClient.identify throws an Exception', () async {
          when(
            () => ldClient.identify(ldContext),
          ).thenThrow(Exception('oops'));

          await expectLater(
            launchDarklyClient.identify(
              userId: userId,
              deviceId: deviceId,
              practiceId: practiceId,
            ),
            throwsA(isA<LaunchDarklyClientIdentificationException>()),
          );
        });

        test('when LDClient.identify returns an IdentifyError', () async {
          when(
            () => ldClient.identify(ldContext),
          ).thenAnswer((_) => Future.value(IdentifyError('oops')));

          await expectLater(
            launchDarklyClient.identify(
              userId: userId,
              deviceId: deviceId,
              practiceId: practiceId,
            ),
            throwsA(isA<LaunchDarklyClientIdentificationException>()),
          );
        });
      });
    });
 });
}
```

The first two tests actually fail (and the latter two pass, but only because they throw an exception for the same reason as the first two tests) because the `LDContext` used in mocking expectations, such as:

```dart
when(
  () => ldClient.identify(ldContext),
).thenAnswer((_) => Future.value(IdentifySuperseded()));
```

cannot be used here as the test does not recognize this as a match against what actually gets called, even though they both construct an `LDContext` with the same data and same method. I'm not sure why they wouldn't be considered equal in this case (which if they were, there wouldn't really be an issue here), but typically how we'd handle this is by providing a fake, such as:

```dart
class _FakeLDContext extends Fake implements LDContext {}
```

so that we could change the above expectation to:

```dart
when(
  () => ldClient.identify(any()),
).thenAnswer((_) => Future.value(IdentifySuperseded()));
```

Any ideas on how this could be achieved?

**Comment by kinyoklion** [comment]
I will need to look into this more deeply, but this doesn't look like a case for mocking. LDContext should be thought of more like a string, or other raw type.

In regards to how to allow `any()` you would need to provide a fallback value for mockito.

```
  setUpAll(() {
    registerFallbackValue(
        LDContextBuilder().kind('not-used', 'not-used').build());
  });
```

That is how you let it have an alternative to a default for non-raw types.

With that added this test will pass:
```
        test('when LDClient.identify returns IdentifyComplete', () async {
          when(
            () => ldClient.identify(any()),
          ).thenAnswer((_) => Future.value(IdentifyComplete()));

          expect(
            () => launchDarklyClient.identify(
              userId: userId,
              deviceId: deviceId,
              practiceId: practiceId,
            ),
            returnsNormally,
          );
        });
```

For more information: https://pub.dev/documentation/mocktail/latest/mocktail/registerFallbackValue.html

One tests still fails, and I will post an update when I understand why.

Thank you,
Ryan

**Comment by danallen88** [comment]
Thanks Ryan! I did find that this helped, and that I could also use this:

```dart
when(
            () => ldClient.identify(any(that: isA<LDContext>())),
          ).thenAnswer((_) => Future.value(IdentifySuperseded()));
```
in cases where it wasn't critical to determine that it's the _same_ `LDContext`. I think that leaves me with the same failing test case as you, in which it would appear that two `LDContext`s constructed in the same way don't appear to be evaluating as equal, but that's just my hunch.

**Comment by kinyoklion** [comment]
> Thanks Ryan! I did find that this helped, and that I could also use this:
> 
> ```dart
> when(
>             () => ldClient.identify(any(that: isA<LDContext>())),
>           ).thenAnswer((_) => Future.value(IdentifySuperseded()));
> ```
> 
> in cases where it wasn't critical to determine that it's the _same_ `LDContext`. I think that leaves me with the same failing test case as you, in which it would appear that two `LDContext`s constructed in the same way don't appear to be evaluating as equal, but that's just my hunch.

We don't provide an equality operator for contexts currently, but we should add it. 

Until then you can do something like this:
```
      test(
        'calls ldClient.identify with userId, practiceId, and deviceId in '
            'context',
            () async {
          when(
                () => ldClient.identify(any()),
          ).thenAnswer((_) => Future.value(IdentifySuperseded()));

          await launchDarklyClient.identify(
            userId: userId,
            deviceId: deviceId,
            practiceId: practiceId,
          );

          final expectedContext = LDContextBuilder()
              .kind('user', userId)
              .kind('device', deviceId)
              .kind('practice', practiceId)
              .build();

          final receivedContext =
          (verify(() => ldClient.identify(captureAny())).captured[0] as LDContext);

          expect(receivedContext.attributesByKind['user']?.key, userId);
          expect(receivedContext.attributesByKind['device']?.key, deviceId);
          expect(receivedContext.attributesByKind['practice']?.key, practiceId);
        },
      );
```

Note I didn't have your kind definitions, so you need to replace 'user', 'device', and 'practice'.

**Comment by danallen88** [comment]
Thank you @kinyoklion, this has been really helpful!

**Comment by kinyoklion** [comment]
@danallen88 Great. Let me know if you need anything else!

**Comment by kinyoklion** [comment]
Adding context equality: https://github.com/launchdarkly/flutter-client-sdk/pull/135

**Comment by kinyoklion** [comment]
@danallen88 Context equality has been released in 4.2.0.

Thank you,
Ryan
